index-scheduler/src/lib.rs

```diff
@@ -1070,7 +1079,24 @@ impl IndexScheduler {
         let batch =
             match self.create_next_batch(&rtxn).map_err(|e| Error::CreateBatch(Box::new(e)))? {
                 Some(batch) => batch,
-                None => return Ok(TickOutcome::WaitForSignal),
+                None => {
+                    // Let's write the previous frame to disk but only if
+                    // the user wanted to profile with puffin.
+                    if puffin_enabled {
+                        let mut frame_view = self.puffin_frame.lock();
+                        if !frame_view.is_empty() {
+                            let now = OffsetDateTime::now_utc();
+                            let mut file = File::create(format!("{}.puffin", now))?;
+                            frame_view.save_to_writer(&mut file)?;
+                            file.sync_all()?;
+                            // We erase this frame view as it is no more useful. We want to
+                            // measure the new frames now that we exported the previous ones.
+                            *frame_view = FrameView::default();
+                        }
+                    }
+
+                    return Ok(TickOutcome::WaitForSignal);
+                }
```

Comment:
Maybe we could move this part in the `run` function to be 100% sure that a batch of tasks has been processed (successful or failed).

Reply:
But we also need to move the puffin frame creation into this run function 🤔 I will try :)